### PR TITLE
Added tapping features

### PR DESCRIPTION
**Pull request by itismeghasyam**
- Added get_tapping_features in get_tapping_features.R
- Added required functions for calculating some tapping features in utils.R
- Added tapping feature extraction related functions in sensors.R

This resolves issue #26

### CODE REVIEW

**Comment by th1vairam** [comment]
@philerooski Check the tapping code from @itismeghasyam. It works perfectly. Just check whether it satisfies the format just like other codes of yours. If so, then finish this merge.

**Comment by itismeghasyam** [comment]
I made most of the changes requested by Phil, set the grounds for casting tapping feature extraction pipeline into a one similar to accelerometer/gyroscope. 

Changes to be made now to incorporate custom feature extraction functions for tapping

**Comment by itismeghasyam** [comment]
Updated tests for added tapping functions. This also resolves issue #20

File: R/sensors.R [hunk 1]
```diff
@@ -27,6 +27,79 @@ sensor_features <- function(sensor_data, transform, extract, extract_on, groups)
...
 }
 
+#' Extract tapping (screen sensor) features
+#' 
+#' @param tap_data A data frame with columns t, x, y, buttonid containing 
+#' tapping measurements. buttonid can be from c('TappedButtonLeft','TappedButtonRight','TappedButtonNone') 
+#' indicating a tap that has been classified as to the left, right or neither of those places on the screen
+#' @param depressThr A numerical threshold for intertap distance in x axis
+#' @return A dataframe of features.
+tapping_features <- function(tap_data,
```
Comment (philerooski) [inline R/sensors.R]: Move immediately before or after `accelerometer_features` and `gyroscope_features`.

File: R/get_tapping_features.R [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+#' indicating a tap that has been classified as to the left, right or neither of those places on the screen
+#' @param depressThr A numeric value indicating the threshold between 
+#' taps (better description needed)
+#' @param removeDups A logical value indicating if duplicates in taps considered as None 
+#' (i.e buttonid is neither left nor right) need to be removed or not
+#' @return Tapping features 
+#' @export
+#' @author Elias Chaibub Neto, Meghasyam Tummalacherla
+get_tapping_features <- function(tap_data, depressThr=20, removeDups=TRUE) {
+  if (is.data.frame(tap_data) == FALSE) {
```
Comment (philerooski) [inline R/get_tapping_features.R]: Remove boolean equality comparisons

File: R/sensors.R [hunk 3]
```diff
@@ -27,6 +27,79 @@ sensor_features <- function(sensor_data, transform, extract, extract_on, groups)
...
+                             depressThr = 20) {
+  
+  results <- GetLeftRightEventsAndTapIntervals(tapData = tap_data,
+                                               depressThr = depressThr)
+  tapInter <- results$tapInter
+  tapData <- results$tapData
+  error <- results$error
+  
+  # check error - if after cleaning tapping data less than 5 data points remain
+  if (error == TRUE) {
```
Comment (philerooski) [inline R/sensors.R]: Remove boolean equality comparisons

File: R/utils.R [hunk 4]
```diff
@@ -1,3 +1,81 @@
+#' Calculate the fatigue given a vector x
+#' 
+#' @param x A numeric vector containing inter tap intevals
+#' @return A list containing fatigue10, fatigue25, fatigue50 where 
+#' fatigueX is the difference in the mean values of the
+#' first X percent of input x and last X percent of input x
+Fatigue <- function(x) {
```
Comment (philerooski) [inline R/utils.R]: lowercase

File: R/utils.R [hunk 5]
```diff
@@ -1,3 +1,81 @@
...
+              fatigue50 = mean(x[1:top50]) - mean(x[(n - top50):n])))
+}
+
+
+#' Calculate the drift given x and y
+#' 
+#' @param x A vecor containing x co-ordinates (same length as that of y)
+#' @param y A vecor containing y co-ordinates (same length as that of x)
+#' @return Drift vector which is sqrt(dx^2 + dy^2)
+calculateDrift <- function(x, y) {
```
Comment (philerooski) [inline R/utils.R]: lower case with underscores

File: R/utils.R [hunk 6]
```diff
@@ -1,3 +1,81 @@
...
+  dy <- diff(y, lag = 1)
+  return(sqrt(dx^2 + dy^2))
+}
+
+
+#' Calculate the Mean Teager-Kaiser energy, adapted from TKEO function in library(seewave) using f = 1, m = 1, M = 1
+#' 
+#' @param x A vector x whose Mean Taiger-Kaiser Energy Operator value needs to be calculated
+#' @return A numeric value that is representative of the MeanTKEO
+MeanTkeo <- function(x) {
```
Comment (philerooski) [inline R/utils.R]: lowercase with underscores

File: R/utils.R [hunk 7]
```diff
@@ -1,3 +1,81 @@
...
+                                         1)])
+  return(mean(y, na.rm = TRUE))
+}
+
+
+#' Calculate the Coefficient of Variation (Cv) for a given sequence
+#' 
+#' @param x A numeric vector x whose Coefficient of Variation needs to be calculated
+#' @return A numeric value that is representative of the Coefficient of Variation
+Cv <- function(x) {
```
Comment (philerooski) [inline R/utils.R]: lowercase

File: R/utils.R [hunk 8]
```diff
@@ -1,3 +1,81 @@
...
+  x <- x[!is.na(x)] # Remove NAs
+  return((sd(x)/mean(x)) * 100)
+}
+
+#' Curate the raw tapping data to get Left and Right events, after applying the threshold
+#' 
+#' @param tapData A dataframe with t,x,y and buttonid columns
+#' @param depressThr The threshold for intertap distance
+#' @return A dataframe with feature values and the appropriate error message
+GetLeftRightEventsAndTapIntervals <- function(tapData, depressThr = 20) {
```
Comment (philerooski) [inline R/utils.R]: lowercase with underscores

File: R/sensors.R [hunk 9]
```diff
@@ -27,6 +27,79 @@ sensor_features <- function(sensor_data, transform, extract, extract_on, groups)
...
+                           tapdata_features)
+  
+  ftrs_error <- grep('error', colnames(tapFeatures))
+  ftrs_error <- paste(tapFeatures[ftrs_error], collapse = ' ; ')
+  if(ftrs_error == ''){
+    ftrs_error = 'None'
+  }
+  
+  tapFeatures$error <- ftrs_error
+  return(tapFeatures %>% as.data.frame())
```
Comment (philerooski) [inline R/sensors.R]: Move the piping function up one line and just return a single object with no "function stuff" within the return statement.

**Comment by itismeghasyam** [comment]
Made the requested changes :)
I think it's better if I put off making any further changes until we close this pull request

**Pull request merged**